File: java/src/main/java/com/sasteval/vuln/medium/SsrfService.java
package com.sasteval.vuln.medium;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

/**
 * CWE-918: Server-Side Request Forgery (SSRF)
 * VULNERABILITY: User-supplied URL is fetched without validation.
 */
public class SsrfService extends HttpServlet {

    private String fetchUrl(String url) throws IOException {
        // VULN: No validation of URL target (allows internal network access)
        HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
        conn.setRequestMethod("GET");
        conn.setConnectTimeout(5000);
        conn.setReadTimeout(5000);

        StringBuilder result = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(conn.getInputStream()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                result.append(line).append("\n");
            }
        }
        return result.toString();
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String url = request.getParameter("url");
        response.setContentType("text/plain");

        try {
            String content = fetchUrl(url);
            response.getWriter().println(content);
        } catch (Exception e) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            response.getWriter().println("Error fetching URL: " + e.getMessage());
        }
    }
}

File: dotnet/Vuln/Medium/SsrfService.cs
namespace SastEval.Vuln.Medium;

/// <summary>
/// CWE-918: Server-Side Request Forgery - User-supplied URL fetched without validation.
/// </summary>
public static class SsrfService
{
    private static async Task<string> FetchUrl(string url)
    {
        // VULNERABLE: No URL validation, allowlist, or blocklist
        using var client = new HttpClient();
        return await client.GetStringAsync(url);
    }

    public static async Task<IResult> Handle(HttpContext context)
    {
        var url = context.Request.Query["url"].ToString();

        if (string.IsNullOrEmpty(url))
        {
            return Results.BadRequest("Missing 'url' query parameter");
        }

        try
        {
            // VULNERABLE: User-controlled URL passed directly to HTTP client
            var content = await FetchUrl(url);
            return Results.Text(content);
        }
        catch (Exception ex)
        {
            return Results.Problem($"Error fetching URL: {ex.Message}");
        }
    }
}
